### PR TITLE
fix fallthrough logic for make_fx(pre_autograd=True)

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/100055

I left more comments in the PR, but there were two issues that both stemmed from "ProxyTorchDispatchMode is supposed effectively be a no-op when it runs re-entrantly" (more details at this note: https://github.com/pytorch/pytorch/blob/a33ac4454058d25abe43532fbe86930e7f55bdda/torch/utils/_python_dispatch.py#L86)

(1) We were (incorrectly) attaching proxies to the input tensors of each op, when the mode was run re-entrantly

(2) We have some logic to return `NotImplemented` if our mode can't handle the inputs it was passed. The linked issue hit a problem where we:

(a) hit `linear()` with ProxyMode
(b) desugared into `addmm` with ProxyMode, with FakeTensor inputs

What we **want** to happen is for the re-entrant ProxyMode call to no-op and redispatch, so we hit FakeTensorMode, and execute `addmm` using FakeTensorMode. Instead, we hit the NotImplemented code path. I didn't dig too much deeper into what exactly it was about hitting the not-implemented path that caused the error, but hitting it went against the idea of "we should no-op when we re-enter ProxyTorchDispatchMode".

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103026
* #103025
* #103024
* #101818
* __->__ #101817

### CODE REVIEW

File: torch/fx/experimental/proxy_tensor.py [hunk 1]
```diff
@@ -258,7 +258,9 @@ def can_handle_tensor(x):
     # If there are any tensor subclasses, we need to handle those tensor subclasses first
     # TODO: we could use types to test this
     if not pytree.tree_all_only(torch.Tensor, can_handle_tensor, (args, kwargs)):
-        return NotImplemented
+        # When is_inside_mode is set, we are logically a fallthrough.
+        if not proxy_mode.is_inside_mode:
+            return NotImplemented
```
Comment (bdhirsh) [inline torch/fx/experimental/proxy_tensor.py:263]: Following the goal of "our mode should be a no-op if it is executed re-entrantly". We have existing logic to raise NotImplemented if we can't handle some inputs.

This feels morally like the right thing: "there are multiple modes but we picked the wrong one, try another!" seems like something that should happen when we invoke ProxyTorchDispatchMode the first time. But on any re-entrant calls to that mode, it should just be a pass through (as if it was never executed).

Comment (ezyang) [reply torch/fx/experimental/proxy_tensor.py:263]: nit: you can test proxy_mode.is_inside_mode first in the outer conditional, saving you a tree_all if it triggers false.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Isn't our long term plan to not reentrantly enter proxy tensor mode, because this is an artifact of the fallthrough handling?

**Comment by ezyang** [comment]
thanks, appreciated.

**Pull request closed without merging**